src/main/java/io/nats/client/impl/VertxDispatcherMessageQueue.java

```diff
@@ -0,0 +1,109 @@
+package io.nats.client.impl;
+
+import io.nats.client.MessageHandler;
+import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+
+import java.time.Duration;
+import java.util.function.Predicate;
+
+public class VertxDispatcherMessageQueue extends MessageQueue {
+    private final VertxDispatcher dispatcher;
+    private final Vertx vertx;
+
+    VertxDispatcherMessageQueue(VertxDispatcher dispatcher, Vertx vertx) {
+        super(true);
+        this.dispatcher = dispatcher;
+        this.vertx = vertx;
+    }
+
+    @Override
+    void pause() {
+        running.set(STOPPED);
+    }
+
+    @Override
+    void drain() {
+        running.set(DRAINING);
+    }
+
+    @Override
+    boolean push(NatsMessage msg) {
+        ((ContextInternal)vertx.getOrCreateContext()).dispatch(e -> {
```

Comment:
As far as I understand `push` will run on a connection thread, so `vertx.getOrCreateContext` will create a new Vertx context for that thread. That is not what we want here probably. Usually we have to capture the Verticle context that did the subscribe and use that particular context to run message handlers.
Additionally `dispatch` assumes the caller thread is the context thread which is not true here, as we are on the connection thread. Instead we probably need `runOnContext` or a variant of it.

Reply:
A message that will be dispatched comes in on the connection's thread.

`vertx` is given to the dispatcher from NatsClientImpl instance, the same place that makes the connection, so they are the same vertx and same context. I added some debug to make sure

```
connect | vertx=io.vertx.core.impl.VertxImpl@59d2400d | context=io.vertx.core.impl.EventLoopContext@75cd8043

push | vertx=io.vertx.core.impl.VertxImpl@59d2400d | context=io.vertx.core.impl.EventLoopContext@70bae6b3
```

So as far as dispatch versus runOnContext, I can change to this:
```
((ContextInternal)vertx.getOrCreateContext()).runOnContext(e -> {
```
instead of this
```
((ContextInternal)vertx.getOrCreateContext()).dispatch(e -> {
```
I can change it. It's interesting. This is the `ContextInternal` implementation of `runOnContext`
```
default void runOnContext(Handler<Void> action) {
  executor().execute(() -> dispatch(action));
}
```
and `dispatch` is this
```
default void dispatch(Handler<Void> handler) {
 dispatch(null, handler);
```